### PR TITLE
Regexdispatcher

### PR DESCRIPTION
**Pull request by e1528532**
# Purpose

Calculates regexes for checker plugins for the use with the typechecker, e.g. converting check/range=0-5000 to a regex representation of that range, also supports the enum checks currently.

# Checklist

- [X] commit messages are fine ("module: short statement" syntax and refer to issues)
- [X] I added unit tests
- [X] I ran all tests locally and everything went fine
- [X] affected documentation is fixed
- [X] I added code comments, logging, and assertions (see doc/CODING.md)
- [X] meta data is updated (e.g. README.md of plugins)
- [x] release notes are updated (doc/news/_preparation_next_release.md)

@markus2330 based on the typechecker branch, the question here is how many things i want to support. I'd add the validation keyword as it more or less represents regexes directly but then lets see, i'd like to get back to writing soon. Basically only the src/plugins/regexdispatcher folder is really relevant and it will be easier to see once the prototype is merged to main.

### CODE REVIEW

**Comment by e1528532** [comment]
jenkins build haskell please

**Comment by e1528532** [comment]
jenkins build haskell please

**Comment by e1528532** [comment]
jenkins build haskell please

**Comment by e1528532** [comment]
jenkins build libelektra please

**Comment by e1528532** [comment]
jenkins build all please

**Comment by e1528532** [comment]
@ingwinlu is there possible some cache involved in the docker container? Can i look into the workspace? This issue 

```
setup: can't find source for Elektra/PluginProcess in

/home/jenkins/workspace/libelektra_PR-1944-NHTCZWJOQXIAMT5YFRVQPHMSUTCZBFPXHGIAGV2EUFI3EU7I6QNQ@2/src/bindings/haskell/src,

dist/build/autogen
```

is not reproducible on my local debian, the initial case issue should be fixed with the commit dcf7995.

**Comment by ingwinlu** [comment]
> is there possible some cache involved in the docker container? 

No, workspace is cleaned before and after every checkout

> Can i look into the workspace? 

The workspace is deleted completely for that test case.

> This issue is not reproducible on my local debian

It is for me. Always fails.

Double check your last commit is it on purpose you not only mo
You have a typo.

Your commit:
haskell/src/Elektra/Pluginprocess.chs → ...ngs/haskell/src/elektra/PluginProcess.chs
The error:
can not find Elektra/PluginProcess

So maybe try renming elektra/PluginProcess.chs to Elektra/PluginProcess.chs.

**Comment by e1528532** [comment]
i see. since i use git on mac that doesn't really differentiate between cases i've totally overlooked that. i guess it was still cached somehow on my debian vm. good catch!

**Comment by e1528532** [comment]
jenkins build all please

**Comment by ingwinlu** [comment]
> good catch!

I will tell the build system the next time i log in.

**Comment by e1528532** [comment]
jenkins build all please

**Comment by e1528532** [comment]
jenkins build gcc-i386 please

**Comment by e1528532** [comment]
jenkins build all please

**Comment by markus2330** [comment]
Sorry, we both fixed the same problem. Please rebase.

**Comment by e1528532** [comment]
jenkins build all please

**Comment by e1528532** [comment]
jenkins build libelektra please

**Comment by e1528532** [comment]
@markus2330 here you go finally

File: src/plugins/regexdispatcher/README.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+- infos = Information about the regexdispatcher plugin is in keys below
+- infos/author = e1528532 <e1528532@libelektra.org>
+- infos/licence = BSD
+- infos/needs =
+- infos/provides = 
+- infos/recommends =
+- infos/placements = presetstorage
+- infos/status = maintained experimental
+- infos/metadata =
```
Comment (markus2330) [inline src/plugins/regexdispatcher/README.md]: Which metadata is used?

File: src/plugins/regexdispatcher/README.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+- infos/description = a plugin which generates regex-representations of specification keywords
+
+## Introduction
+
+A plugin which generates regex-representations of specification keywords for the
```
Comment (markus2330) [inline src/plugins/regexdispatcher/README.md]: Input/output is unclear: which meta data is consumed, which is generated?

File: src/plugins/regexdispatcher/README.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+- infos/placements = presetstorage
+- infos/status = maintained experimental
+- infos/metadata =
+- infos/description = a plugin which generates regex-representations of specification keywords
+
+## Introduction
+
+A plugin which generates regex-representations of specification keywords for the 
+typechecker plugin before saving a configuration specification. Intended to be 
+mounted before the typechecker plugin.
```
Comment (markus2330) [inline src/plugins/regexdispatcher/README.md]: "before" is misleading: the actual order need to be specified with placements or infos/ordering

File: src/plugins/regexdispatcher/README.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+## Usage
+
+In order to generate regex representations for different specification keywords, mount a configuration specification along with this plugin. 
+
+`kdb mount <specification> spec/<path> <storage plugin to read the specification> regexdispatcher typechecker`
```
Comment (markus2330) [inline src/plugins/regexdispatcher/README.md]: Why not let the typechecker depend on the regexdispatcher? Or is it usful to use the typechecker without regexdispatcher!

Please describe more clearly why the regexdispatcher is needed.

File: src/plugins/regexdispatcher/README.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+## Dependencies
+
+* ghc >= 8.0.1 < 8.4
+* ghc-pkg, usually bundled with ghc
+* cabal, the haskell build system, usually bundled with ghc
+* augeas, which provides libfa utilized by this plugin
+
+## Limitations
```
Comment (markus2330) [inline src/plugins/regexdispatcher/README.md]: Wrong section name, you describe features and not limitations here.

File: src/bindings/haskell/src/Elektra/Invoke.chs [hunk 6]
```diff
@@ -0,0 +1,62 @@
+--
+-- @file
+--
+-- @brief invoke Haskell bindings
```
Comment (markus2330) [inline src/bindings/haskell/src/Elektra/Invoke.chs]: "Haskell bindings for libelektra-invoke" is much less confusing

File: src/libs/pluginprocess/pluginprocess.c [hunk 7]
```diff
@@ -401,3 +403,34 @@ int elektraPluginProcessClose (ElektraPluginProcess * pp, Key * errorKey)
 	if (done) cleanupPluginData (pp, errorKey);
 	return done;
 }
+
+/** Store a pointer to any plugin related data that is being executed inside an own process.
+ *
+ * @param plugin a pointer to the plugin
+ * @param data the pointer to the data
+ */
+void elektraPluginProcessSetData (Plugin * handle, void * data)
```
Comment (markus2330) [inline src/libs/pluginprocess/pluginprocess.c]: How is this serialized and sent to the other process?

File: src/plugins/haskell/haskell.c.in [hunk 8]
```diff
@@ -16,6 +16,11 @@
 #include <kdbpluginprocess.h>
 #include <stdio.h>
 
+// TODO:
```
Comment (markus2330) [inline src/plugins/haskell/haskell.c.in]: Thank you for writing the TODO. Having a way to store plugin data is essential.

File: src/plugins/regexdispatcher/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,9 @@
+include (LibAddHaskellPlugin)
+
+find_package (fa)
+
+if (LIBFA_FOUND)
+	add_haskell_plugin (regexdispatcher MODULE "RegexDispatcher" SANDBOX_ADD_SOURCES "src/libs/typesystem/libfa")
```
Comment (markus2330) [inline src/plugins/regexdispatcher/CMakeLists.txt]: how are the tests specified?

File: src/plugins/regexdispatcher/Elektra/Parsers.hs [hunk 10]
```diff
@@ -0,0 +1,44 @@
+--
+-- @file
+--
+-- @brief Some parsers which are used in the translation process
```
Comment (markus2330) [inline src/plugins/regexdispatcher/Elektra/Parsers.hs]: Not clear to me what the parsers are for.

File: src/plugins/regexdispatcher/README.md [hunk 11]
```diff
@@ -0,0 +1,36 @@
+- infos = Information about the regexdispatcher plugin is in keys below
+- infos/author = e1528532 <e1528532@libelektra.org>
```
Comment (markus2330) [inline src/plugins/regexdispatcher/README.md]: Please write your name.

File: src/plugins/regexdispatcher/README.md (hunk 4 above)
Comment (e1528532) [reply src/plugins/regexdispatcher/README.md]: i will. basically the typechecker can also work without the regexdispatcher for certain metadata like check/long or check/validation or similar which do not need an intermediate translation process.
However something like check/enum/#1=A, check/enum/#2 = B needs to be translated to the regex A|B, which the regexdispatcher handles to keep concerns separated.
Furthermore other people may implement similar plugins that translate their own metakeys to regexes (that don't necessarily have to be implemented in haskell) and so i concluded to keep those things separated from the typechecker plugin.

File: src/plugins/haskell/haskell.c.in (hunk 8 above)
Comment (e1528532) [reply src/plugins/haskell/haskell.c.in]: actually i've forgotten to remove this TODO, this is already implemented as i needed it to pass the linux builds.

File: src/plugins/regexdispatcher/CMakeLists.txt (hunk 9 above)
Comment (e1528532) [reply src/plugins/regexdispatcher/CMakeLists.txt]: The tests are written in C

File: src/plugins/regexdispatcher/README.md (hunk 4 above)
Comment (markus2330) [reply src/plugins/regexdispatcher/README.md]: Yes, I assumed something like that. But the docu does not say it.

File: src/plugins/haskell/haskell.c.in (hunk 8 above)
Comment (markus2330) [reply src/plugins/haskell/haskell.c.in]: Even better :+1:

**Comment by e1528532** [comment]
jenkins build all please

**Comment by e1528532** [comment]
some very good points you've raised there especially regarding the ordering and the metakeys and the purpose.

File: doc/METADATA.ini [hunk 12]
```diff
@@ -905,3 +905,13 @@ default = dec
 usedby/plugins = hexnumber
 description = used to specify the base of an integer value. Currently only the hexnumber plugin supports this metadata. It converts any value marked as "hex" from
     hexadecimal into decimal when the configuration is read, and back when it is written. The hexnumber plugin also ensures that a value marked as "hex" starts with "0x" or "0X".
+
+[elektra/spec/regex/check/enum]
+type = string
+usedby/plugins = regexdispatcher
+description = the generated regex representing the type of check/enum/# and check/enum/multi metakeys
+
+[elektra/spec/regex/check/range]
```
Comment (markus2330) [inline doc/METADATA.ini]: Some thoughts about the metadata name:

- the elektra prefix does not give information, all meta data is for and from Elektra.
- it should also not start with spec (do not be confused with the namespace).

And wouldn't it be better that we only have a single metadata and the regex is always merged together? Do you need the information of where the regex comes from?

What was the idea of giving this name?

Comment (e1528532) [reply doc/METADATA.ini]: you are right. as the regexdispatcher does all the preprocessing this essentially yields a check/validation keyword in the end all the time, so in that case i don't need further information where it came from i think. i will try that idea out.

**Comment by e1528532** [comment]
jenkins build all please

**Comment by e1528532** [comment]
jenkins build all please

**Comment by e1528532** [comment]
@markus2330 as discussed in #160 these few checks are unlikely to every succeed and thus i'd suggest merging it now? as far as i know they don't check any haskell related things anyway.

**Comment by markus2330** [comment]
thank you!

**Pull request merged**